### PR TITLE
[EXPERIMENTAL] A new flat API without a breaking change

### PR DESCRIPTION
**Pull request by suevalov**
An experimental branch to create a new optional flat API which would simplify building web applications.

The naming is subject to change:
```javascript
import { createFlatClient } from 'contentful-managemnt';

const client = createFlatClient({ accessToken: '' }, { 
   spaceId: 'my-space',
   environmentId: 'master', // defaults are specified during initialization
});

let entry = await client.entry.get({ entryId: 'jkahdksjhkjhas' });
entry.fields.name['en-US'] = 'new-name';
entry = await client.entry.update({ entryId: entry.sys.id }, entry);

const space = await client.space.get(); // gets a default space
const differentSpace = await client.space.get({ spaceId: 'space-different-from-my-default-settings'):

const environment = await client.environment.get(); // gets a default env
const differentEnvironment = await client.environment.get({ environmentId: 'environment-different-from-my-default-settings'):
```

The most important that the new client is opt-in and we are not doing a breaking change.

```javascript
// updating entry with a current sdk

const space = await client.getSpace(spaceId);
const environment = await space.getEnvironment(envId);
let entry = await environment.getEntry(entryId);
entry = await entry.update(newData)
```

### CODE REVIEW

File: lib/plain/plain-client.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { AxiosInstance } from 'axios'
+
+const withHttp = <T extends any[], R>(
+  http: AxiosInstance,
+  fn: (http: AxiosInstance, ...rest: T) => R
+) => {
+  return (...rest: T) => fn(http, ...rest)
+}
+
+export const createPlainClient = (params: ClientParams) => {
```
Comment (anho) [inline lib/plain/plain-client.ts:12]: What about:

```

type EndpointMethod<T extends any[], R> = (http: AxiosInstance, ...rest: T) => R;
type EndpointMethodDict<T extends any[], R> = Record<string, EndpointMethod<T, R>>;

const withHttp = <T extends any[], R>(http: AxiosInstance, fn: EndpointMethod<T, R>) => (...rest: T) => fn(http, ...rest)

const fromEndpoints = <T extends any[], R>(http: AxiosInstance, endpoints: EndpointMethodDict<T, R>) => {
  return Object.entries(endpoints)
    .reduce(
      (endpointsWithHttp, [methodName, method]) => ({ ...endpointsWithHttp, [methodName]: withHttp(http, method) }),
      {},
    );
};

export const createPlainClient = (params: ClientParams) => {
  const http = createCMAHttpClient(params)

  return {
    space: fromEndpoints(http, endpoints.space),
    environment: fromEndpoints(http, endpoints.environment),
    contentType: fromEndpoints(http, endpoints.contentType),
    user: fromEndpoints(http, endpoints.user),
    entry: fromEndpoints(http, endpoints.entry),
  }
}
```

Comment (suevalov) [reply lib/plain/plain-client.ts:12]: Awesome if all types work as expected. I'll check it out.

File: lib/plain/index.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        params: params.query,
+      }
+    )
+  },
+}
+
+export const user = {
+  getManyForSpace(http: AxiosInstance, params: { spaceId: string; query?: QueryOptions }) {
+    return get<CollectionProp<UserProps>>(http, `/spaces/${params.spaceId}/users`, {
+      params: params.query,
```
[the remark this answers is not in the document]
Comment (suevalov) [reply lib/plain/index.ts:120]: `normalizeSelect` is only for several queries right now, it's a good question if we need to add it everywhere or not

File: lib/plain/index.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    })
+  },
+}
+
+export const entry = {
+  getMany(
```
[the remark this answers is not in the document]
Comment (suevalov) [reply lib/plain/index.ts:126]: I'm not sure if `normalizeSelect` is needed everywhere.

**Comment by suevalov** [comment]
I'm going to merge it to `experimental` branch from which we automatically release a new experimental version.

**Pull request merged**